### ISSUE TITLE
Data channels: support unreliable communication

### ISSUE THREAD

**Issue by backkem** [pions/webrtc#334]
Support unreliable data channels.
Needed by [decentraland/communications-server-go](https://github.com/decentraland/communications-server-go).

**Comment by enobufs** [comment]
@backkem This issue has become urgent to my project. Could you tell me how far this has been addressed? If not much, I could take a look at it this weekend. If there's any blocker to this issue you are aware of, please let me know.

**Comment by enobufs** [comment]
These params...
```
	MaxPacketLifeTime *uint16
	MaxRetransmits *uint16
```
being pointers look a bit odd to me. I know spec says about "null", but it also says:

> If a setting, either [[MaxPacketLifeTime]] or [[MaxRetransmits]], has been set to indicate unreliable mode, **and that value exceeds the maximum value supported by the user agent, the value MUST be set to the user agents maximum value**.

If the max value for pion (the "user agent") can be less than 2^31 (I belive reasonable values are much less than that!), then we could use `int16`, so that we can consider negative values as "null" for JavaScript... my 2c.

**Comment by enobufs** [comment]
[RFC3758](https://tools.ietf.org/html/rfc3758) "SCTP Partial Reliability Extension", this is what we need to implement. Description for how to give up retransmission is not defined in this spec.

**Comment by enobufs** [comment]
Ok, that's [DCEP])https://tools.ietf.org/html/draft-ietf-rtcweb-data-protocol-09), I guess...

**Comment by enobufs** [comment]
This is much harder than I thought (^^;
From my notes:
* PR-SCTP needs to be implemented (RFC 3758)
* Enable other ChannelTypes.
  - Need a new arg to NewRTCDataChannel() for the channel type (currenly hard-coded)
 - Uncomment code implemented earlier that is to determine datachannel params in pc.CreateDataChannel().
* DTLSRole matters to determine Stream Identifier, or SI according to DCEP spec, to avoid SI collision...
  - RTCDataChannel.ID seems to become SI...  data channel ID == stream ID ?

Any suggestion to how we should approach?

**Comment by backkem** [comment]
- Regarding the params: They are uint16 since that's what the [spec](https://w3c.github.io/webrtc-pc/#dom-rtcdatachannelinit) says. It also spec says that most of these these parameters are optional. They are pointers in `RTCDataChannelInit` to be able to tell the difference between them being set and not being set.  Otherwise we can't tell the difference between `RTCDataChannelInit{}` and `RTCDataChannelInit{MaxPacketLifeTime:0}`. We already had a rather lengthy discussion about this in the past: close adherence to the spec vs go-like vs .... Allowing negative values seems equally ugly to me. `webrtc/pkg/null` was another idea we explored. I'm not sure if we want to open that discussion again. I do agree that `RTCDataChannel.MaxPacketLifeTime` has no reason being a 
pointer. Also, I would be in favor of replacing these attributes if `RTCDataChannel` with methods so we can control thread safety.
- Regarding the configuration: What we are doing now is we have the WebRTC API as a wrapper around the ORTC API. So the code that is commented out in `CreateDataChannel` needs to take the input from the [RTCDataChannelInit](https://w3c.github.io/webrtc-pc/#dom-rtcdatachannelinit) and supply it to `NewRTCDataChannel` using the [RTCDataChannelParameters](https://draft.ortc.org/#dom-rtcdatachannelparameters). The `RTCDataChannelInit` has all attributes implemented (#107). `RTCDataChannelParameters` only has the attributes implemented that actually do something. The rest still has to be added.
- We currently generate the datachannel IDs in `RTCPeerConnection.generateDataChannelID`. The user also has the option to supply their own ID.

I'll also start reading up on on the protocol specs again.

**Comment by backkem** [comment]
High level breakdown:
- Implement the Forward-TSN-Supported Parameter and FORWARD TSN chunk.
- Implement the PR-SCTP logic. May require implementing the T3-rtx timer.
- Extend the SCTP API to configure the reliability. The spec suggests doing this per message. In that case I would extend the WriteSCTP to something like:
```go
type WriteInfo struct {
  PPI PayloadProtocolIdentifier
  Reliable bool
  Ordered bool
  TTL time.Duration
}
(s *Stream) WriteSCTP(b []byte, p WriteInfo) (n int, err error)
```
- Extend the DataChannel API: I would set the reliabilty per channel and use that to populate the `WriteInfo`s used to call `WriteSCTP`. Note: `All Data Channel Establishment Protocol messages MUST be sent using ordered delivery and reliable transmission.`
- ORTC API (mentioned above): Extend `RTCDataChannelParameters` and forward to datachannel.
- WebRTC API (mentioned above): Forward `RTCDataChannelInit` to `RTCDataChannelParameters`.

**Comment by enobufs** [comment]
Thanks @backkem , your comments are really helpful.
So, I guess current user data retransmission relies only on acks from the remote - which could rather be a bigger problem but we could add the t3-rtx timer altogether in this work.

**Comment by enobufs** [comment]
Hi @backkem I am still working on this issue... Could you tell me what [this flag](https://github.com/pions/sctp/blob/7c3fee5926cd5ed551d38f3c6bbd5bec1e405bbe/chunk_payload_data.go#L63) is for?
```
const (
	payloadDataEndingFragmentBitmask   = 1
	payloadDataBeginingFragmentBitmask = 2
	payloadDataUnorderedBitmask        = 4
	payloadDataImmediateSACK           = 8 // <----- this one
```
I could not find any info about it in any of RFC 2960, 4960 and PR-SCTP, and it doesn't appear to be really used.  I could remove it if no longer needed.

**Comment by enobufs** [comment]
Here's [my note](https://docs.google.com/document/d/1i4FgCd0gZc1aHh7zcA_m81yIzvsxHXejTMKpcu0SM4s/edit?usp=sharing).

**Comment by enobufs** [comment]
Before implementing FORWARD TSN, I had to sort out things around reassebly queue.  Issues I came across were descripbed in my note (see my previous message). Those are all fixed in rev 29698. Improvement made so far are:

* Unordered option works now (was not fully implemented. Added tests also)
* Ordered stream is no longer blocked by another stream of different SI.

I am about to start implementing Forward TSN - finally.

**Comment by enobufs** [comment]
Using webrtc@sctp-pr branch (at rev@fcdfe), I was able to connect to Chrome and data going back and forth with no problem. I can see from the console output (from sctp) that it is correctly taking reliability parameters from API, and the biggest thing is Chrome accepts all permutations of Order, MaxPacketLifeTime and MaxRetransmits with no problem (whew..)
I am going to try to cause packet loss (maybe using XCode's Network Link Conditioner) to see of Foward TSN works ok with Chrome.

**Comment by enobufs** [comment]
With MaxRetransmits set to 1 with Chrome, and Network Link Conditioner with 10% packet loss ratio, I can see Forward TSN chunks are going back and forth. I can see from the log that some packets are abandoned "correctly" and TSNs are moving on:
```
Message 'Payload Type String' from DataChannel 'data' payload 'Message 10'
Sending Message 12
final (abandoned) tsn=3790043284 (remix: 1)
sending tsn=3790043284 ssn=12 sent=1
building fwdtsn: si=0 ssn=12 tsn=3790043284 acked=false
stream[0]: handleData ts=2488767947 ssn=11
stream[0]: handleData readable? true
discard tsn=2488767947
Message 'Payload Type String' from DataChannel 'data' payload 'Message 11'
building fwdtsn: si=0 ssn=12 tsn=3790043284 acked=false
Sending Message 13
final (abandoned) tsn=3790043285 (remix: 1)
sending tsn=3790043285 ssn=13 sent=1
building fwdtsn: si=0 ssn=13 tsn=3790043285 acked=false
building fwdtsn: si=0 ssn=13 tsn=3790043285 acked=false
Sending Message 14
final (abandoned) tsn=3790043286 (remix: 1)
sending tsn=3790043286 ssn=14 sent=1
building fwdtsn: si=0 ssn=14 tsn=3790043286 acked=false
Sending Message 15
final (abandoned) tsn=3790043287 (remix: 1)
sending tsn=3790043287 ssn=15 sent=1
building fwdtsn: si=0 ssn=15 tsn=3790043287 acked=false
stream[0]: handleData ts=2488767948 ssn=12
stream[0]: handleData readable? true
discard tsn=2488767948
Message 'Payload Type String' from DataChannel 'data' payload 'Message 14'
```
I believe Chrome is happy with FORWARD TSN chunk pion generates. :)

I also see this line many times:
```
building fwdtsn: si=0 ssn=13 tsn=3790043285 acked=false
```
The line would show up when retransmission occurred. With the lack of T3-rtx timer, we are sending all unacked chunks every time SACK arrives.  We will need to fix this as soon as this work is complete.

**Comment by enobufs** [comment]
I also confirmed that the accepting peer (Chrome)'s RTCDataChannel object has the same reliability parameters as what initiator requested:
```
ordered: true
maxPacketLifeTime: undefined
maxRetransmits: 1
```
FYI.

**Comment by enobufs** [comment]
Following modules have changes related to this issue:
* pions/transport@sctp-pr
* pions/sctp@sctp-pr (depends on transport@sctp-r)
* pions/datachannel@sctp-pr (depends on transport@sctp-r and sctp@sctp-pr)
* pions/webrtc@issue-334r (depends on transport@sctp-pr, sctp@sctp-pr and datachannel@sctp-pr)

To consolidate go.mod, we will need to release in the following order:
1. Release transport with a tag
1. Release sctp with a tag
1. Release datachannel with a tag
1. Finally, release webrtc with a tag
